### PR TITLE
ggml: parallelize dequantization or fp format conversion when using blas

### LINKED ISSUE #4988: Parallelize dequantization or fp format conversion when using blas
In ggml_compute_forward_mul_mat@ggml.c, gemm parallelism is managed by blas library. However, this disables multithreading on dequantizing weights, which may be a bottleneck.

I have performed some ugly modifications for comparing performance.

```
› MKL_NUM_THREADS=8 ./llama-bench -m Llama-2-7b-chat-q4km.gguf -t 8
| model                          |       size |     params | backend    |    threads | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ---------: | ---------- | ---------------: |
| llama 7B Q4_K - Medium         |   3.80 GiB |     6.74 B | BLAS       |          8 | pp 512     |     55.24 ± 3.16 |
| llama 7B Q4_K - Medium         |   3.80 GiB |     6.74 B | BLAS       |          8 | tg 128     |     21.12 ± 1.89 |

build: 862f5e4 (1886)

› MKL_NUM_THREADS=8 ./llama-bench -m Llama-2-7b-chat-q4km.gguf -t 8
| model                          |       size |     params | backend    |    threads | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | ---------: | ---------- | ---------------: |
| llama 7B Q4_K - Medium         |   3.80 GiB |     6.74 B | BLAS       |          8 | pp 512     |     79.34 ± 5.23 |
| llama 7B Q4_K - Medium         |   3.80 GiB |     6.74 B | BLAS       |          8 | tg 128     |     21.26 ± 1.83 |

modified
```
[cb828c8](https://github.com/ReinForce-II/llama.cpp/commit/cb828c8d5a1851f0c9ab5b8c1609ff1ebd761fdc)

### PR DESCRIPTION
**Pull request by ReinForce-II**
resolve #4988 

* make GGML_TASK_INIT phase can be run in multithread
* move dequantization in mul_mat to GGML_TASK_INIT phase when using BLAS lib and enable multithreading
* update work size to fit fully dequantized weights

### CODE REVIEW

**Comment by slaren** [comment]
This is good as it is, but in the future I think we could remove the init and finalize tasks, and instead have a barrier synchronization function that waits until all the threads working on an op reach the same point. Then the ops could have as many phases as they need, and it might simplify the implementation a bit.

File: ggml.c [hunk 1]
```diff
@@ -16696,7 +16797,9 @@ struct ggml_cplan ggml_graph_plan(const struct ggml_cgraph * cgraph, int n_threa
                     if (ggml_compute_forward_mul_mat_use_blas(node)) {
                         if (node->src[0]->type != GGML_TYPE_F32) {
                             // here we need memory just for single 2D matrix from src0
```
Comment (ggerganov) [inline ggml.c:16799]: This comment is no longer correct - should be updated

File: ggml.c [hunk 2]
```diff
@@ -9825,21 +9828,45 @@ static void ggml_compute_forward_mul_mat(
...
+                    for (int64_t i12 = 0; i12 < ne12; i12++) {
+                        // broadcast src0 into src1 across 2nd,3rd dimension
+                        const int64_t i03 = i13/r3;
+                        const int64_t i02 = i12/r2;
+
+                        const void           *       x        = (char *)  src0->data    + i02*nb02          + i03*nb03;
+                              float          * const wdata    = (float *) params->wdata + i13*ne12*ne_plane + i12*ne_plane;
+                              ggml_to_float_t  const to_float = type_traits[type].to_float;
+
+                        for (int64_t i01 = ith; i01 < ne01; i01+=nth) {
```
Comment (ggerganov) [inline ggml.c:9849]:
```suggestion
                        for (int64_t i01 = ith; i01 < ne01; i01 += nth) {
```

**Review by ggerganov — APPROVED** [review]
I'm not convinced about the `sched_yield` changes - on macOS I do see performance improvements with CPU + Accelerate, but on my Linux machine with OpenBLAS, `whisper.cpp` is slower with this change.

I guess we will merge it and look into a more general solution of removing the tasks and synchronizing on barriers

**Comment by ReinForce-II** [comment]
On Linux with OpenBLAS, when it's running with all available threads by default, a spin not managed in blas lib can compete with it, which can lead to significant downgraded performance. Num threads used by openblas can be set via utility functions in openblas extension, but an equivlant function is not always available in general blas providers. I guess the case can be relieved by environment variable, but it still need something better here.

File: ggml.c [hunk 3]
```diff
@@ -16695,8 +16796,10 @@ struct ggml_cplan ggml_graph_plan(const struct ggml_cgraph * cgraph, int n_threa
 #if defined(GGML_USE_ACCELERATE) || defined(GGML_USE_OPENBLAS)
                     if (ggml_compute_forward_mul_mat_use_blas(node)) {
                         if (node->src[0]->type != GGML_TYPE_F32) {
-                            // here we need memory just for single 2D matrix from src0
-                            cur = ggml_type_size(GGML_TYPE_F32)*(node->src[0]->ne[0]*node->src[0]->ne[1]);
+                            // here we need memory for fully dequantized matrix from src0
+                            cur = ggml_type_size(GGML_TYPE_F32)
+                                * node->src[0]->ne[0]*node->src[0]->ne[1]
+                                * node->src[1]->ne[2]*node->src[1]->ne[3];
```
Comment (ggerganov) [inline ggml.c:16802]: Can be simplified to:

```suggestion
                            cur = ggml_type_size(GGML_TYPE_F32)*ggml_nelements(node->src[0]);
```

Comment (ggerganov) [reply ggml.c]: Nevermind - it has to be like this to take into account the broadcast

Comment (cebtenzzre) [reply ggml.c]: @ggerganov it looks like this suggestion was merged, but it sounds like you didn't mean for it to be?

Comment (ggerganov) [reply ggml.c]: Fixed here: https://github.com/ggerganov/llama.cpp/pull/5094

**Comment by JohannesGaessler** [comment]
This PR seems to have caused a large performance regression for at least one user: https://github.com/ggerganov/llama.cpp/issues/6417

**Pull request merged**